chrome/content/zotero/xpcom/connector/server_connector.js

```diff
@@ -384,6 +384,70 @@ Zotero.Server.Connector.SaveSession.prototype._updateRecents = function () {
 };
 
 
+Zotero.Server.Connector.Utilities = {
+
+	/**
+	 * Helper function to insert form data back into SingleFileZ pageData object
+	 *
+	 * SingleFileZ creates a single object containing all page data including all
+	 * resource files. We turn that into a multipart/form-data request for upload
+	 * and here we insert the form resources back into the SingleFileZ object.
+	 *
+	 * @param {Object} resources - Resources object inside SingleFileZ pageData object
+	 * @param {Object} formData - Multipart form data as a keyed object
+	 */
+	insertSnapshotResources: function (resources, formData) {
+		for (let resourceType in resources) {
+			for (let resource of resources[resourceType]) {
+				// Frames have whole new set of resources
+				// We handle these by recursion
+				if (resourceType === "frames") {
+					Zotero.Server.Connector.Utilities.insertSnapshotResources(resource.resources, formData);
+					return;
+				}
+				// UUIDs are 8 characters in length
+				if (resource.content.length === 8) {
+					// Replace content with actual content indexed in formData
+					// by the UUID stored in the content
+					resource.content = formData.find(
+						element => element.headers.name === resource.content
+					).body;
+				}
+			}
+		}
+	},
+	
+	/**
+	 * Helper function to decode a raw JS string
+	 *
+	 * When the server retrieves a multipart request it doesn't know which parts are binary and
+	 * which are UTF-8 encoded strings. So, after we have removed all binary data, we need to
+	 * do the decoding on our own.
+	 *
+	 * @param {String} data - Raw input string
+	 * @return {String} String decoded using UTF-8
+	 */
+	decodeRawString: function (data) {
+		// Convert to UTF-8
+		let dataStream = Components.classes["@mozilla.org/io/string-input-stream;1"]
+					.createInstance(Components.interfaces.nsIStringInputStream);
+		dataStream.setData(data, data.length);
+
+		let utf8Stream = Components.classes["@mozilla.org/intl/converter-input-stream;1"]
+					.createInstance(Components.interfaces.nsIConverterInputStream);
+		utf8Stream.init(dataStream, "UTF-8", 4096, "?");
+
+		let decoded = "";
+		let string = {};
+		while (utf8Stream.readString(data.length, string)) {
+			decoded += string.value;
+		}
```

Comment:
I know this is copied from `_decodeRawData()`, but these days we can do this in a more standards-compliant fashion with TextDecoder:

```javascript
var buf = new ArrayBuffer(data.length);
var bufView = new Uint8Array(buf);
for (let i = 0; i < data.length; i++) {
	bufView[i] = data.charCodeAt(i);
}
var decoder = new TextDecoder();
return decoder.decode(bufView);
```

(Not fully tested, but something like that should work.)

Probably better to add that to a Utilities.Internal.decodeUTF8(data) function (maybe above `byteLength()`) that we can just call from both of these places.